File: vane-core/src/main/java/org/oddlama/vane/util/LazyBlock.java
package org.oddlama.vane.util;

import java.util.UUID;
import org.bukkit.Bukkit;
import org.bukkit.block.Block;

public class LazyBlock {

    private final UUID worldId;
    private int x;
    private int y;
    private int z;
    private Block block;

    public LazyBlock(final Block block) {
        if (block == null) {
            this.worldId = null;
            this.x = 0;
            this.y = 0;
            this.z = 0;
        } else {
            this.worldId = block.getWorld().getUID();
            this.x = block.getX();
            this.y = block.getY();
            this.z = block.getZ();
        }
        this.block = block;
    }

    public LazyBlock(final UUID worldId, int x, int y, int z) {
        this.worldId = worldId;
        this.x = x;
        this.y = y;
        this.z = z;
        this.block = null;
    }

    public UUID worldId() {
        return worldId;
    }

    public int x() {
        return x;
    }

    public int y() {
        return y;
    }

    public int z() {
        return z;
    }

    public Block block() {
        if (worldId != null && block == null) {
            this.block = Bukkit.getWorld(worldId).getBlockAt(x, y, z);
        }

        return block;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/util/LazyBlock.kt
package org.oddlama.vane.util

import org.bukkit.Bukkit
import org.bukkit.block.Block
import java.util.*

/**
 * Lazily resolves block references from stored world UUID and coordinates.
 *
 * @param worldId world UUID used for lazy world lookup.
 * @param x block x coordinate.
 * @param y block y coordinate.
 * @param z block z coordinate.
 */
class LazyBlock(
    /** World UUID used for lazy resolution. */
    val worldId: UUID?,
    /** Block x coordinate. */
    val x: Int,
    /** Block y coordinate. */
    val y: Int,
    /** Block z coordinate. */
    val z: Int,
) {
    /**
     * Cached resolved block instance.
     */
    private var block: Block? = null

    /**
     * Creates a lazy block from an existing block reference.
     */
    constructor(block: Block?) : this(
        worldId = block?.world?.uid,
        x = block?.x ?: 0,
        y = block?.y ?: 0,
        z = block?.z ?: 0,
    ) {
        this.block = block
    }

    /**
     * Returns the resolved block, loading world/block lazily when needed.
     */
    fun block(): Block? {
        if (worldId != null && block == null) {
            block = Bukkit.getWorld(worldId)!!.getBlockAt(x, y, z)
        }
        return block
    }
}
